src/_pytest/logging.py

```diff
@@ -614,27 +614,17 @@ def pytest_collection(self) -> Generator[None, None, None]:
             with catching_logs(self.log_file_handler, level=self.log_file_level):
                 yield
 
-    def _runtest_for(
-        self, item: Optional[nodes.Item], when: str
-    ) -> Generator[None, None, None]:
+    def _runtest_for(self, item: nodes.Item, when: str) -> Generator[None, None, None]:
         """Implements the internals of pytest_runtest_xxx() hook."""
         log_handler = LogCaptureHandler()
         log_handler.setFormatter(self.formatter)
         with catching_logs(log_handler, level=self.log_level):
-            if item is not None:
-                empty = {}  # type: Dict[str, LogCaptureHandler]
-                item._store.setdefault(catch_log_handlers_key, empty)[
-                    when
-                ] = log_handler
-                item._store[catch_log_handler_key] = log_handler
-            try:
-                yield
-            finally:
-                if item is not None and when == "teardown":
-                    del item._store[catch_log_handlers_key]
-                    del item._store[catch_log_handler_key]
+            item._store[catch_log_handlers_key][when] = log_handler
+            item._store[catch_log_handler_key] = log_handler
+
+            yield
```

Comment:
`try/finally` probably required here?

Reply:
The `yield` can't throw (pluggy only `send`s a Result in), so it's not really needed.